controllers/token_scorer/token_scorer.py

```diff
@@ -0,0 +1,257 @@
+        print(  # noqa:T001
+            f"{token_code} NOW WORTH {token_value} TO {token_code.owner.name} "
+            f"AT {claim_time}s",
+        )
+
+    def record_captures(self) -> None:
+        if not self._record_arena_actions:
+            self._log_is_dirty = False  # Stop links and displays being updated every timestep
+            return
+
+        if not self._log_is_dirty:
+            # Don't write the log if nothing new has happened.
+            return
+
+        controller_utils.record_arena_data({'token_claims': [
+            {
+                'zone': claim.token_code.owner.value,
+                'token_index': claim.token_code.id,
+                'token_value': claim.token_value,
+                'time': claim.claim_time,
+            }
+            for claim in self._log
+        ]})
+
+        self._log_is_dirty = False
+
+    def is_dirty(self) -> bool:
+        return self._log_is_dirty
+
+
+def order_zone_points(zone: Tuple[Point, Point]) -> Tuple[Point, Point]:
+    'Sort points so the first point has the lower x and y coordinates'
+    return (
+        Point(min(zone[0].x, zone[1].x), min(zone[0].y, zone[1].y)),
+        Point(max(zone[0].x, zone[1].x), max(zone[0].y, zone[1].y)),
+    )
+
+
+class TokenScorer:
+    def __init__(
+        self,
+        claim_log: ClaimLog,
+        ship_zone: Tuple[Point, Point],
+        zone_0_stack: Tuple[Point, Point],
+        zone_1_stack: Tuple[Point, Point],
+    ) -> None:
+        self._claim_log = claim_log
+        self._robot = Supervisor()
+        self.ship_zone = order_zone_points(ship_zone)
+        self.stack_zones = [order_zone_points(zone_0_stack), order_zone_points(zone_1_stack)]
+
+        self._token_statuses: Dict[Token, int] = {
+            code: 0 for code in TOKENS
+        }
+
+        self._scoring_receivers: Dict[Receiver, Point] = {
+            get_robot_device(self._robot, 'token scorer', Receiver): Point(0, -1.25),
+        }
+
+        timestep = int(self._robot.getBasicTimeStep())
+        for receiver in self._scoring_receivers.keys():
+            receiver.enable(timestep)
+
+    def token_in_zone(self, token_location: Point, zone: Tuple[Point, Point]) -> bool:
+        # zone[0] < zone[1]
+        if token_location.x < zone[0].x or zone[1].x < token_location.x:
+            # token is outside the zone in the x-direction
+            return False
+        if token_location.y < zone[0].y or zone[1].y < token_location.y:
+            # token is outside the zone in the x-direction
+            return False
+
+        return True
+
+    def get_token_value(
+        self,
+        owner: Owner,
+        location: Point,
+    ) -> int:
+        value = 0
+
+        if self.token_in_zone(location, self.ship_zone):
+            value = FLOOR_TOKEN_VALUE
+
+        if self.token_in_zone(location, self.stack_zones[owner]):
+            value += PLATFORM_EXTRA_TOKEN_VALUE
+
+        return value
+
+    def get_token_location(
```

Comment:
Meta: Sounds like there's no way to strictly enumerate the tokens and get their positions directly through the supervisor API?

Reply:
Previously when the tokens were freestanding, you could use the supervisor API to directly read the `translation` field. However this required a large number of SWIG calls so performance was poor. Additionally since we are now using the token harness this data is no longer stored in a single field and crosses PROTO boundaries.